aten/src/ATen/core/dynamic_type.h

```diff
@@ -0,0 +1,118 @@
+
+#include <ATen/core/class_type.h>
+#include <ATen/core/jit_type_base.h>
+#include <c10/util/Optional.h>
+
+using DynamicTypeBits = std::uint32_t;
+#define DYNAMIC_TYPE_BIT(x) (1u << x)
+
+constexpr DynamicTypeBits kDynamicCovariantTypeBit = DYNAMIC_TYPE_BIT(31);
+constexpr DynamicTypeBits kDynamicAnyTypeBit = DYNAMIC_TYPE_BIT(30);
+constexpr DynamicTypeBits kDynamicSingletonTypeBit = DYNAMIC_TYPE_BIT(29);
+
+constexpr DynamicTypeBits kDynamicNoneTypeBit = DYNAMIC_TYPE_BIT(1);
+constexpr DynamicTypeBits kDynamicIntTypeBit = DYNAMIC_TYPE_BIT(3);
+constexpr DynamicTypeBits kDynamicFloatTypeBit = DYNAMIC_TYPE_BIT(4);
+constexpr DynamicTypeBits kDynamicComplexTypeBit = DYNAMIC_TYPE_BIT(5);
+constexpr DynamicTypeBits kDynamicListTypeBit = DYNAMIC_TYPE_BIT(7);
+constexpr DynamicTypeBits kDynamicTupleTypeBit = DYNAMIC_TYPE_BIT(8);
+
+
+#define FORALL_DYNAMIC_TYPES(_)                                              \
+  _(Tensor, DYNAMIC_TYPE_BIT(0))                                             \
+  _(None, kDynamicNoneTypeBit)                                               \
+  _(Bool, DYNAMIC_TYPE_BIT(2))                                               \
+  _(Int, kDynamicIntTypeBit)                                                 \
+  _(Float, kDynamicFloatTypeBit)                                             \
+  _(Complex, kDynamicComplexTypeBit)                                         \
+  _(Number,                                                                  \
+    (kDynamicIntTypeBit | kDynamicFloatTypeBit | kDynamicComplexTypeBit))    \
+  _(String, DYNAMIC_TYPE_BIT(6))                                             \
+  _(List, kDynamicListTypeBit)                                               \
+  _(Tuple, (kDynamicTupleTypeBit | kDynamicCovariantTypeBit))                \
+  _(Dict, DYNAMIC_TYPE_BIT(9))                                               \
+  _(Class, DYNAMIC_TYPE_BIT(10))                                             \
+  _(Optional,                                                                \
+    (DYNAMIC_TYPE_BIT(11) | kDynamicNoneTypeBit | kDynamicCovariantTypeBit)) \
+  _(AnyList, (kDynamicListTypeBit | kDynamicAnyTypeBit))                     \
+  _(AnyTuple,                                                                \
+    (kDynamicTupleTypeBit | kDynamicCovariantTypeBit | kDynamicAnyTypeBit))  \
+  _(Any, 0xffffffff)
+
+namespace c10 {
+
+struct FunctionSchema;
+struct LabeledDynamicType;
+class DynamicType;
+using DynamicTypePtr = std::shared_ptr<DynamicType>;
+
+// Low dependency jit type with minimal subtyping and structucing support,
+// designed for embedded cases.
+class DynamicType : public Type {
+  using ClassTypePtr = std::shared_ptr<const c10::ClassType>;
+
+ public:
+  ~DynamicType() override;
+
+  struct Arguments {
+    Arguments() = default;
+    Arguments(c10::ArrayRef<TypePtr>);
+    Arguments(const c10::FunctionSchema&, c10::ArrayRef<TypePtr>);
+    std::vector<LabeledDynamicType> elems;
+  };
```

Comment:
yes, contained types. FunctionSchema is here because I need the argument names. I should just change it to `std::vector<std::string>` to avoid confusion.